proposals/0010-resource-element-type-validation.md

```diff
@@ -0,0 +1,223 @@
+  * [RW]Buffer
+  * [Feedback]Texture*
+
+There is a distinct set of rules that define valid RETs for raw buffer types, 
+and a separate set of rules that define valid RETs for typed buffer types.
+These rules also depend on the target IR, SPIR-V or DXIL.
+
+RETs for typed buffer variants with a DXIL target IR may include:
+* basic types: 
+  * 16, 32, and 64-bit int and uint
+  * half, float, and double
+* vectors and matrices 
+  * containing 4 elements or fewer
+  * total size may not exceed 128 bits
+* user defined types (structs / classes), as long as:
+  * all fields in the struct have the same type
+  * there are at most 4 sub elements
+  * total size may not exceed 128 bits
+
+RETs for raw buffer variants are much less constrained:
+* it must be a complete type
+* cannot contain a handle (such as resource types)
+
+Resource types are never allowed as RETs (i.e., `RWBuffer<int>` as an RET).
+Texture resources conform to the rules for typed buffers.
+If the target is SPIR-V, then only the set of rules for raw buffers apply. Typed buffers
+like `RWBuffer` may have RETs that exceed 128 bits, for example, if the target 
+IR is SPIR-V. The typed buffer rules above are only enforced when the IR target is DXIL.
+
+If someone writes `RWBuffer<MyCustomType>` and MyCustomType is not a 
+valid RET, there should be infrastructure to reject this RET and emit a message 
+explaining why it was rejected as an RET.
+
+## Motivation
+Currently, there is an allow list of valid RETs. It must be modified with respect 
+to this spec. Anything that is not a valid RET will be rejected. The allow list isn't
+broad enough, because it doesn't include the case where the RET is user-defined. 
+Ideally, a user should be able to determine how any user-defined structure is invalid 
+as an RET. Some system should be in place to more completely enforce the rules for 
+valid and invalid RETs, as well as provide useful information on why they are invalid.
+
+For example, when targeting DXIL IR, `RWBuffer<double4> b : register(u4);` will emit
+an error in DXC, but will not in clang-dxc, despite the fact that `double4` is an 
+invalid RET for typed buffers.
+
+## Proposed solution
+
+The proposed solution is to use some type_traits defined in the std library, create
+some custom builtins to use as type_traits, and join them together to define a 
+set of conceptual constraints for any RET that is used. These conceptual constraints
+will be applied to every typed buffer resource type that is defined, so that all
+typed buffer HLSL resources have the same rules about which RETs are valid. 
+Validation will occur upon resource type instantiation. Additionally, certain 
+resource types are raw buffer variants, such as `StructuredBuffer`. These resource 
+types will have a different set of type-traits applied, which will loosen constraints
+on viable RETs. The type traits will be validated as `concepts`, which is a feature 
+introduced in C++20 to statically validate a type.

```

Comment:
I think this paragraph can be simplified somewhat, I suspect it has evolved into it's current form.

```suggestion
The proposed solution is to use C++20 concepts to validate the element type.
A new built-in, `__builtin_is_homogenous`, will be added in order to express constraints
that can't currently be expressed in pure HLSL code. Standard clang diagnostics for 
unsatisfied constrints will be used to report any invalid element types.

Until concepts are formally supported by HLSL, the concepts and constraints 
will be expressed directly in the AST via the HLSL external sema source.
```